chain/client/src/stateless_validation/chunk_endorsement_tracker.rs

```diff
@@ -40,33 +55,66 @@ impl ChunkEndorsementTracker {
         Self {
             epoch_manager,
             chunk_endorsements: SyncLruCache::new(NUM_CHUNKS_IN_CHUNK_ENDORSEMENTS_CACHE),
+            // We can use a different cache size if needed, it does not have to be the same as for `chunk_endorsements`.
+            pending_chunk_endorsements: SyncLruCache::new(NUM_CHUNKS_IN_CHUNK_ENDORSEMENTS_CACHE),
         }
     }
 
+    /// Process pending endorsements for the given chunk header.
+    /// It removes these endorsements from the `pending_chunk_endorsements` cache.
+    pub fn process_pending_endorsements(
+        &self,
+        chunk_header: &ShardChunkHeader,
+    ) -> Result<(), Error> {
+        let chunk_hash = &chunk_header.chunk_hash();
+        let chunk_endorsements = {
+            let mut guard = self.pending_chunk_endorsements.lock();
+            guard.pop(chunk_hash)
+        };
+        let chunk_endorsements = match chunk_endorsements {
+            Some(chunk_endorsements) => chunk_endorsements,
+            None => {
+                tracing::debug!(target: "stateless_validation", ?chunk_hash, "No pending chunk endorsements.");
+                return Ok(());
+            }
+        };
+        chunk_endorsements.values().try_for_each(|endorsement| {
+            self.process_chunk_endorsement(endorsement.clone(), Some(&chunk_header))
+        })
+    }
+
     /// Function to process an incoming chunk endorsement from chunk validators.
-    /// We first verify the chunk endorsement and then store it in a cache.
+    /// If the chunk header is available, we will verify the chunk endorsement and then store it in a cache.
+    /// Otherwise, we store the endorsement in a separate cache of endorsements to be processed when the chunk is ready.
     /// We would later include the endorsements in the block production.
     pub(crate) fn process_chunk_endorsement(
         &self,
-        chunk_header: &ShardChunkHeader,
         endorsement: ChunkEndorsement,
+        chunk_header: Option<&ShardChunkHeader>,
```

Comment:
I personally would have been a tiny bit happier if we had two separate (internal) functions to deal with adding chunk endorsement to normal vs pending cache. It gets a bit confusing understanding the code in terms of "ok, wait, when can I have an optional header? What are the consequences of not passing the header?". 

`process_chunk_endorsement` takes non-optional chunk_header
`add_chunk_endorsement_to_pending_cache` takes just endorsement with no chunk_header

I'll leave the implementation detail decision to you :)

Reply:
Honestly it's even fine if we still have two separate functions and both call some `process_chunk_endorsement_impl` function 😅